File: dashboardview/src/main/java/cn/cxy/dashboardview/MainActivity.java
package cn.cxy.dashboardview;

import android.os.Bundle;
import android.widget.SeekBar;

import androidx.appcompat.app.AppCompatActivity;

import cn.cxy.dashboardview.view.DashboardView;

public class MainActivity extends AppCompatActivity {
    DashboardView dashboardView;
    SeekBar seekBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        dashboardView = (DashboardView) findViewById(R.id.dashboardView);
        seekBar = (SeekBar) findViewById(R.id.seekBar);
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                dashboardView.setPercent(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
    }
}

File: stars/src/main/java/cn/cxy/stars/MainActivity.kt
package cn.cxy.stars

import android.os.Bundle
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    private var starWidth = 10
    private var startHeight = 10

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        updateChildViews(rootLayout)
        rootLayout.setOnClickListener { updateChildViews(rootLayout) }
    }

    private fun updateChildViews(viewGroup: ViewGroup) {
        viewGroup.removeAllViews()
        viewGroup.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                viewGroup.viewTreeObserver.removeOnPreDrawListener(this)
                (0..100).forEach {
                    addStarView(getRandomRatio(), getRandomRatio(), viewGroup.width, viewGroup.height, starWidth, startHeight)
                }
                return true
            }
        })
    }

    private fun getRandomRatio() = ((1..100).random() / 100.toFloat())

    private fun addStarView(x: Float, y: Float, parentWidth: Int, parentHeight: Int, width: Int, height: Int) {
        val view = LightSpotViewWithAnimation(this)
        val layoutParams = FrameLayout.LayoutParams(dp2Px(width), dp2Px(height))
        layoutParams.leftMargin = (x * parentWidth).toInt()
        layoutParams.topMargin = (y * parentHeight).toInt()
        rootLayout.addView(view, layoutParams)
    }

    fun dp2Px(dpValue: Int) = dp2Px(this, dpValue)
}